src/assignments/tech-jobs-oo.rst

```diff
@@ -1,6 +1,586 @@
+--------------------------------------------
+
+To display the data for a particular ``Job`` object, you need to implement a
+custom ``toString`` method. Rather than creating this method and then testing
+it, you will flip that process using TDD.
+
+Create First Test for ``toString``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Before writing your first test, consider how we want the method to behave:
+
+#. When passed a ``Job`` object, it should return a string that contains a
+   blank line before and after the job information.
+#. The string should contain a label for each field, followed by the data
+   stored in that field. Each field should be on its own line.
+
+   .. sourcecode:: bash
+
+      ID:  _______
+      Name: _______
+      Employer: _______
+      Location: _______
+      Position Type: _______
+      Core Competency: _______
+
+#. If a field is empty, the method should add, "Data not available" after
+   the label.
+#. (Bonus) If a ``Job`` object ONLY contains data for the ``id`` field, the
+   method should return, "OOPS! This job does not seem to exist."
+
+In ``JobClassTests``, add a new test to check the first requirement, then run
+that test (it should fail).
+
+Woo hoo! Failure is what we want here! Now you get to fix that.
+
+Code ``toString`` to Pass the First Test
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the ``Job`` class, create a ``toString`` method that passes the first test.
+Since the test only checks if the returned string starts and ends with a blank
+line, make that happen.
+
+.. admonition:: Tip
+
+   Do NOT add anything beyond what is needed to make the test pass. You will
+   add the remaining behaviors for ``toString`` as you code each new test.
+
+Finish the TDD for ``toString``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+#. Code a new test for the second required behavior, then run the tests to make
+   sure the new one fails.
+#. Modify ``toString`` to make the new test pass. Also, make sure that your
+   updates still pass all of the old tests.
+#. Continue this test-refactor cycle until all of the behaviors we want for
+   ``toString`` work. Remember to add only ONE new test at a time.
+
+Cool! Your ``Job`` class is now complete and operates as desired.
+
+6) Refactor to DRY the Support Classes
+---------------------------------------
+
+Review the code in the ``Employer``, ``Location``, ``CoreCompetency``, and
+``PositionType`` classes. What similarities do you see?
+
+There is a fair amount of repetition between the classes. As a good coder,
+anytime you find yourself adding identical code in multiple locations you
+should consider how to streamline the process.
+
+   DRY = "Don't Repeat Yourself".
+
+Create a Base Class
+^^^^^^^^^^^^^^^^^^^^
+
+Let's move all of the repeated code into a separate class. We will then have
+``Employer``, ``Location``, ``CoreCompetency``, and ``PositionType`` *inherit*
+this common code.
+
+#. Create a new class called ``JobFields``.
```

Comment:
Class names should be singular, in most situations: `JobField`.

Reply:
More than one field, hence ``JobFields``.  Is this the exception to the "most situations" clause?

Reply:
Fixed.